core/blockchain/impl/block_tree_impl.cpp

```diff
@@ -853,34 +857,42 @@ namespace kagome::blockchain {
 
   outcome::result<void> BlockTreeImpl::prune(
       const std::shared_ptr<TreeNode> &lastFinalizedNode) {
-    std::vector<std::pair<primitives::BlockHash, primitives::BlockNumber>>
-        to_remove;
-
-    auto current_node = lastFinalizedNode;
-
-    for (auto parent_node = current_node->parent.lock();
-         parent_node && !parent_node->finalized;
-         parent_node = current_node->parent.lock()) {
-      auto main_chain_node = current_node;
-      current_node = parent_node;
-
-      // collect hashes for removing (except main chain block)
-      for (const auto &child : current_node->children) {
-        if (child->block_hash != main_chain_node->block_hash) {
-          collectDescendants(child, to_remove);
-          to_remove.emplace_back(child->block_hash, child->depth);
-        }
+    std::deque<std::shared_ptr<TreeNode>> to_remove;
+
+    auto following_node = lastFinalizedNode;
+
+    for (auto current_node = following_node->parent.lock();
+         current_node
+         && !(current_node->finalized && current_node->children.size() == 1);
+         current_node = current_node->parent.lock()) {
+      // DFS-on-deque
+      to_remove.emplace_back();  // Waterbreak
+      std::copy_if(
+          current_node->children.begin(),
+          current_node->children.end(),
+          std::back_inserter(to_remove),
+          [&following_node](auto child) { return child != following_node; });
+      auto last = to_remove.back();
+      while (last != nullptr) {
+        to_remove.pop_back();
+        std::copy(last->children.begin(),
+                  last->children.end(),
+                  std::back_inserter(to_remove));
+        to_remove.emplace_front(std::move(last));
+        last = to_remove.back();
```

Comment:
Hm... I think std::rotate is the best algorithm  for that

Reply:
std::rotate for deque is linear. These should work in constant time